### ISSUE TITLE
Implement ObjectMaterialized event

### ISSUE THREAD

**Issue by ajcvickers** [dotnet/efcore#15911]
Splitting this out as a separate issue as required on #626 so we can track it independently of other life-cycle hooks.

The `ObjectMaterialized` event will fire after an entity instance has been created and all it's non-navigation properties have been set.

**Comment by weitzhandler** [comment]
`ObjectMaterializing` as well, to be able to set some default values or process candidate data?

**Comment by SidShetye** [comment]
@ajcvickers thanks for breaking this down into even smaller work items. Wasn't sure what granularity @divega was looking for.

Anyway, could this be promoted to 3.0.0 milestone just like https://github.com/aspnet/EntityFrameworkCore/issues/15910 ? The reason being we need hooks at both paths to and from the database - to preserve symmetry and avoid data corruption during data transformations. Having both `ObjectMaterialized` and `SavingChanges` is the bare minimum for us to unblock our EF Core customers who've been patiently asking us to support EF Core for a very long time. At our end, we can spotlight the 3.0.0 release as "fully supported" when it eventually comes out. 

Thanks

**Comment by ajcvickers** [comment]
@SidShetye Can you give some details on how you intend to use `ObjectMaterialized`?

**Comment by SidShetye** [comment]
@ajcvickers : Sure. We basically use integration points in any ORM to enforce our data security pipeline. 

Specifically, for data about to be written to the database we'd use something like `SavingChanges` in EF that results in encrypted data being sent to the database. 

So on the reverse path, we again need our security pipeline to kick in that results in plaintext data being presented to the EF application. `ObjectMaterialized` in EF 6.x allows us that attach point and we'd like to use it similarly in EF Core too. As you can imagine, without anything like `ObjectMaterialized` on the reverse path, the data remains in an unusable state. For a visual, check out the very first figure on [page 1](https://patentimages.storage.googleapis.com/a6/74/b4/9dbd9d94313619/US20150339486A1.pdf).

**Comment by SidShetye** [comment]
@ajcvickers - will this be marked for the 3.0 release?

**Comment by ajcvickers** [comment]
@SidShetye It's pretty unlikely to go into 3.0 at this point.

**Comment by SidShetye** [comment]
@ajcvickers Could we please have this for 3.1?

**Comment by optiks** [comment]
@SidShetye Have you considered using value conversions instead? Presumably you have a mechanism for flagging which properties need to be encrypted? 

You could also consider introducing an Encrypted<T> (potentially with an implicit operator) and expose that in your domain.

https://docs.microsoft.com/en-us/ef/core/modeling/value-conversions

**Comment by ajcvickers** [comment]
@SidShetye We haven't decided what will be in 3.1 yet.

**Comment by nphmuller** [comment]
With this event we could easily implement child entity ordering, which would be great to have.

**Comment by EMooreMAC** [comment]
We need this feature because our entities track their own changes independently of the DbContext, and our framework needs to know the difference between a property being set on initialization vs later. In EF6, we can set an IsInitializing flag to false after the object is materialized, and our tracking code (in property setters) can ignore property changes that happen during initialization.

I'm open to ideas as to how else we could achieve similar functionality.

**Comment by optiks** [comment]
> I'm open to ideas as to how else we could achieve similar functionality.

@EMooreMAC, I’ve solved this in the past. Give me a few days and I’ll try and get some code for you.

**Comment by StevenRasmussen** [comment]
@optiks - I'm looking for something similar.  Have you had a chance to pull that code together by chance?

**Comment by StevenRasmussen** [comment]
Ok.... posting this for anyone else that would like a workaround in the short term until this is implemented.  This isn't fully tested so use at your own risk... and with the caveat that the underlying code *may* change at any time since it is relying on some EF Core internals. Perhaps someone from the EF team could comment on how stable/usable this might be for the short term. That being said... I'm using the code with great success.

Declare an interface for methods to be invoked
```csharp
interface IMaterialize
{
    void OnMaterializing();
    void OnMaterialized();
}
```

Implement the interface on your model:
```csharp
class YourModel : IMaterialize
{
    public void OnMaterializing()
    {
        // Called after the constructor has been called and object created
    }
    public void OnMaterialized()
    {
        // Called after all properties have been populated
    }
}
```
Create a new implementation of the `EntityMaterializerSource` class which inherits from the EFCore implementation but overrides one of the methods:
```csharp
class MyEntityMaterializerSource : Microsoft.EntityFrameworkCore.Query.EntityMaterializerSource
{
    public MyEntityMaterializerSource(EntityMaterializerSourceDependencies dependencies)
        : base(dependencies)
    {
    }

    public override Expression CreateMaterializeExpression(IEntityType entityType, string entityInstanceName, Expression materializationContextExpression)
    {
        var baseExpression = base.CreateMaterializeExpression(entityType, entityInstanceName, materializationContextExpression);
        if (entityType.ClrType.GetInterfaces().FirstOrDefault(i => i == typeof(IMaterialize)) != null)
        {
            var onMaterializingMethod = entityType.ClrType.GetMethod(nameof(IMaterialize.OnMaterializing));
            var onMaterializedMethod = entityType.ClrType.GetMethod(nameof(IMaterialize.OnMaterialized));

            var blockExpressions = new List<Expression>(((BlockExpression)baseExpression).Expressions);
            var instanceVariable = blockExpressions.Last() as ParameterExpression;

            var onMaterializingExpression = Expression.Call(instanceVariable, onMaterializingMethod);
            var onMaterializedExpression = Expression.Call(instanceVariable, onMaterializedMethod);

            blockExpressions.Insert(1, onMaterializingExpression);
            blockExpressions.Insert(blockExpressions.Count - 1, onMaterializedExpression);

            return Expression.Block(new[] { instanceVariable }, blockExpressions);
        }

        return baseExpression;
    }
}
```

Now simply replace the service while adding the DbContext to the service collection or in the `Dbcontext.OnConfiguring`:

```csharp
services.AddDbContext<TestContext>(options =>
{
    options.ReplaceService<IEntityMaterializerSource, MyEntityMaterializerSource>();
}...

// Or...
protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
{
    optionsBuilder.ReplaceService<IEntityMaterializerSource, MyEntityMaterializerSource>();
}
```

**Comment by aktxyz** [comment]
wow ... MyEntityMaterializerSource works great :)  an official approach will be great but for now this is perfect

**Comment by titobf** [comment]
@StevenRasmussen thanks so much for that workaround! I have just tried it and it works!

**Comment by haiplk** [comment]
Hi all,,
Just curious about the override EntityMaterializerSource cannot work with nested select

`dbContext.Table1.Select(f=> new CustomModel{  DateTimeField = f.DateTime }).ToListAsync()`

Do you all have any idea on it

**Comment by EMooreMAC** [comment]
@StevenRasmussen just wanted to check back and say that this solution is working for us as well. Thanks a ton, you helped clear a blockage for our large scale migration to EF Core.

**Comment by smitpatel** [comment]
@haiplk - Your query does not materialize any entity.

**Comment by Plasma** [comment]
https://github.com/dotnet/efcore/issues/15911#issuecomment-613174136 Is a good solution, but a proper callback method being called (eg to provide an object authorisation service audit the object) would be ideal.

Are there any plans for this? Thank you

**Comment by r-work** [comment]
@StevenRasmussen thanks for the workaround, however `OnMaterialized` gets called as soon as the "parent" entity is materialized before its eager loaded sub collections are loaded/materialized.

So unfortunately this won't work if you want a callback after the "whole" entity is loaded.

**Comment by avenmore** [comment]
It also seems that the implementation in [#15911 (comment)](https://github.com/dotnet/efcore/issues/15911#issuecomment-613174136) does not call `OnMaterialized()` for entities fetched with `.Include()`. Also wish for a real solution.

**Comment by Plasma** [comment]
I was reflecting on how this event would work, the proposed event name is OnMaterialize, is the intention it would run only once an object is materialised, or also if the object was in query cache and is being returned from a query?

My personal use case for this kind of event is to have a more extensive query filter (that can filter or throw an exception) if the entity being returned is not valid for the current security context.

1) Would the event run every time, even if the entity already exists in query cache?

2) Alternatively, can the global Query Filter system support client side filtering too

3) Or, is there a way to add a filter on top of DbSet/Set responses (can’t seem to figure it out).

Thank you!

**Comment by avenmore** [comment]
I missed this event when porting some code from EF6/Wpf where a base-class property `Materialised` was always set on all entities and this helped preventing premature INotifyPropertyChanged events when it hadn't been set. I never gave much thought as to how EF6 implemented it, just that it worked really well - perhaps that can serve as a spec.

**Comment by ajcvickers** [comment]
Note from triage: consider configure await when implementing this.

**Comment by VahidN** [comment]
Is it possible to access the related `DbCommand` [here](https://github.com/dotnet/efcore/pull/28274) as well? It will help to write a good caching interceptor based on this new feature.

**Comment by ajcvickers** [comment]
@VahidN No; this is operating at a different level in the stack.

**Comment by maxime-poulain** [comment]
Hello,

thanks for this, it is exactly what I needed for my current development.(I am updating the `IsTransient` property of my entities thanks to this).
I would like just to add that the [official documentation](https://learn.microsoft.com/en-us/ef/core/logging-events-diagnostics/interceptors) is not mentionning that this is possible thanks to `IMaterializationInterceptor`.

You can find that information however in the following article: [What's new in EF 7](https://learn.microsoft.com/en-us/ef/core/logging-events-diagnostics/interceptors).

But I guess the above link won't be the one people looking for that kind of behavior where they will be looking at.

If you could update the documentation, I am sure it would be helpfull for many people.

**Comment by yokovaski** [comment]
The official documentation is indeed lacking a reference to the new `IMaterializationInterceptor`. [Here](https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-7.0/whatsnew#simple-actions-on-entity-creation) is the direct link to the announcement with an example of how to use it.
